src/core/dom/srcset.js

```diff
@@ -90,31 +88,37 @@ export function parseSrcset(s) {
  * "select" method for details on how this selection is performed.
  *
  * See https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img#Attributes
+ * @template {SrcsetSourceDef} SourceDef
  */
 export class Srcset {
   /**
-   * @param {!Array<!SrcsetSourceDef>} sources
+   * @param {Array<SourceDef>} sources
    */
   constructor(sources) {
     userAssert(sources.length > 0, 'Srcset must have at least one source');
-    /** @private @const {!Array<!SrcsetSourceDef>} */
+    /** @private @const {Array<SourceDef>} */
     this.sources_ = sources;
 
     // Only one type of source specified can be used - width or DPR.
     let hasWidth = false;
     let hasDpr = false;
     for (let i = 0; i < sources.length; i++) {
+      /** @type {?} */
       const source = sources[i];
       hasWidth = hasWidth || !!source.width;
       hasDpr = hasDpr || !!source.dpr;
     }
     userAssert(
-      !!(hasWidth ^ hasDpr),
```

Comment:
Can you explain this line?

Reply:
`^` is disallowed for booleans. JS doesn't perform boolean XOR, it casts each operand to a number, XORs the numbers, and casts the result to a boolean.

Reply:
I just tested this because I'd never seen it before...it actually works!